src/main/java/bio/terra/service/dataset/DatasetRequestValidator.java

```diff
@@ -220,6 +225,20 @@ private void validateTable(TableModel table, Errors errors, SchemaValidationCont
         }
     }
 
+    private void validateDataTypes(List<ColumnModel> columns, Errors errors) {
+        List validDataTypes = new ArrayList(Arrays.asList(TableDatatypes.values()));
+        for (ColumnModel column : columns) {
+            // spring defaults user input not belonging to the TableDataTypes enum to null
+            if (column.getDatatype() == null || !validDataTypes.contains(column.getDatatype())) {
```

Comment:
It's interesting that you get to this point with a null. I would expect the request to be rejected before it reaches this point, since the json data couldn't be marshaled into the java class.

Also, is there a need to check `validDataTypes.contains()` here? What cases will this catch that the JSON -> object conversion isn't already catching?

Reply:
I included this so we don't have to rely on the incidental behavior of spring converting invalid enums to null. It looks like it converts the value to null and adds a message like this to the Errors object before we even reach this DatasetRequesetValidator class
`"message":"Validation errors - see error details","errorDetail":["schema.tables[0].columns[2].datatype: 'NotNull' (must not be null)"`
It then marshalls the json with null as the value. It's not behavior I expected, and I couldn't find any documentation on it.

Reply:
Actually, researching further this conversion to null is not built in spring behavior, it seems like we have a custom converter somewhere?
see: https://stackoverflow.com/questions/39774427/springs-requestparam-with-enum/39774700

Reply:
this as well: https://stackoverflow.com/questions/62774975/springboot-rest-validation-does-not-fail-on-wrong-enum-input
Seems like we are modifying the behavior somewhere. I kind of prefer the null value though, since it lets us send this custom error message that lists all the valid types.

Reply:
But since it looks like it either converts to null or rejects the request, I can remove the `validDataTypes.contains()`

Reply:
It looks like there's a swagger codegen option that controls this behavior (`errorOnUnknownEnum=true`) but I couldn't get it to work. The PR for the feature is here https://github.com/swagger-api/swagger-codegen/pull/10356 and a discussion of the feature is here https://github.com/swagger-api/swagger-codegen/issues/5950. A comment at the bottom of the issue shows how to override the enum serialization behavior without changing the codegen.

The main thing we lose by reporting it at this point is we don't have the original user's source so we can't tell them the string that was wrong. On the other hand, maybe doing more is an over-engineered fix.. Either way the problematic data will be caught before the job is created.